app/Platform/Actions/UnlockPlayerAchievementAction.php

```diff
@@ -4,126 +4,85 @@
 
 namespace App\Platform\Actions;
 
-use App\Platform\Events\AchievementSetCompleted;
 use App\Platform\Events\PlayerAchievementUnlocked;
 use App\Platform\Models\Achievement;
-use App\Platform\Models\Game;
 use App\Platform\Models\PlayerAchievement;
 use App\Site\Models\User;
 use Carbon\Carbon;
+use Exception;
 
 class UnlockPlayerAchievementAction
 {
-    public function execute(User $user, Achievement $achievement, bool $hardcore, ?User $unlockedBy = null): array
-    {
-        // TODO refactor to new schema
+    public function execute(
+        User $user,
+        Achievement $achievement,
+        bool $hardcore,
+        ?Carbon $timestamp = null,
+        ?User $unlockedBy = null
+    ): void {
+        $timestamp ??= Carbon::now();
+
+        if (!$achievement->game) {
+            throw new Exception('Achievement does not belong to any game');
+        }
+
+        // make sure to resume the player session which will attach the game to the player, too
+        $achievement->loadMissing('game');
```

Comment:
```
$ php artisan ra:platform:player:unlock-achievement Jamiras 174647,174648,174649,174650 --hardcore
Unlocking 4 [hardcore] Achievements for user [Jamiras]
 0/4 [░░░░░░░░░░░░░░░░░░░░░░░░░░░░]   0%
   Illuminate\Database\LazyLoadingViolationException 

  Attempted to lazy load [game] on model [App\Platform\Models\Achievement] but lazy loading is disabled.
```

Reply:
This is being caused by this conditional in _UnlockPlayerAchievement.php_:

```php
if (!$achievement->game) {
    throw new Exception('Achievement does not belong to any game');
}
```

The game relationship is being accessed, but the query to load the game hasn't been executed yet. Beginning from Laravel 8, it appears that lazy loading is disabled by default to prevent N+1 query problems.

There is a `loadMissing()` call which happens right after the conditional which, if placed before the conditional, resolves the problem.

From an architectural standpoint, I am unsure if lazy or eager loading is best within the context of this action. It seems like eager loading could be more beneficial here because we know in advance we're going to access this related model, and eager loading reduces the count of queries made to the database. Given this particular action, it seems like the game relationship is crucial for the action to proceed.

I think we could eager load the game by adding this conditional to the very top of the action:

```php
if (!$achievement->relationLoaded('game')) {
  $achievement->load('game');
}
```

Reply:
> The game relationship is being accessed, but the query to load the game hasn't been executed yet. Beginning from Laravel 8, it appears that lazy loading is disabled by default to prevent N+1 query problems.

The forced eager loading behaviour is explicitly enabled here: https://github.com/RetroAchievements/RAWeb/blob/master/app/Site/AppServiceProvider.php#L48

`loadMissing()` does pretty much the same as checking for `relationLoaded` manually. 
I moved the `loadMissing()` call above the check.

Reply:
I can confirm this is now working for me.